File: dannybrown/2022/src/day7/day7.py
LARGEST_SIZE = 100000
MAX_SPACE = 70000000
SPACE_NEEDED = 30000000
class TreeNode:
    def __init__(self, text, size=0, children=[], parent=None):
        self.text = text
        self.size = size
        self.children = children
        self.parent = parent
    def __str__(self):
        return self.text

    def __repr__(self):
        return self.text

def get_full_path(node):
    # the same directory name can appear multiple times with a different full path
    if not node:
        return ""
    return node.text + "/" + get_full_path(node.parent)

def part1_2(node, dictionary):
    val = node.size + sum([part1_2(child, dictionary) for child in node.children])
    if node.children != [] and node.size == 0:
        path_to_node = get_full_path(node)
        dictionary[path_to_node] = val
    return val

if __name__ == '__main__':
    with open('dannybrown/2022/input/day7.txt', 'r') as f:
        lines = f.readlines()

    root = TreeNode(text="root", size=0, children=[], parent=None)

    lines = lines[1:] # always start with cd root
    dir_dict = {}
    current_node = root
    for line in lines:
        line = line.strip()
        if line.startswith("$"):
            # command
            line = line.split(" ")
            command = line[1]
            if command == "cd":
                # breakpoint()
                if line[2] == "..":
                    current_node = current_node.parent
                    continue
                for child in current_node.children:
                    if child.text == line[2]:
                        current_node = child
                        break
            else: #ls
                pass
        else:
            # result of ls
            line = line.split(" ")
            size=0
            if line[0] != "dir":
                size = int(line[0])
            else:
                dir_dict[line[1]] = 0
            text = line[1]
            new_node = TreeNode(text=text, size=size, children=[], parent=current_node)
            current_node.children.append(new_node)

    part1_sum = 0
    dct = {}
    part1_2(root, dct)

    for key, value in dct.items():
        if value < LARGEST_SIZE:
            part1_sum += value
    print('Part 1:', part1_sum)

    # Part 2

    root_size = dct["root/"]
    used_space = MAX_SPACE - root_size

    dirs_to_delete = []
    for key,value in dct.items():
        if used_space + value >= SPACE_NEEDED:
            dirs_to_delete.append(value)

    print('Part 2:', min(dirs_to_delete))

File: dannybrown/2022/src/day7/day7.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const SIZE_TO_CHECK = 100000

func check(e error) {
	if e != nil {
		panic(e)
	}
}

type Node struct {
    text     string
	size     int
    children []*Node
	parent   *Node
}

func main() {
	dat, err := os.Open("dannybrown/2022/input/day7.txt")
	check(err)
	defer dat.Close()
	scanner := bufio.NewScanner(dat)
	scanner.Split(bufio.ScanLines)
	scanner.Scan() // Skip first line cd /
	// input := scanner.Text()
	root := Node{
		text:     "/",
		size:     0,
		children: []*Node{},
		parent:  nil,
	}
	curNode := &root
	for scanner.Scan() {
		input := scanner.Text()
		fmt.Println(input)
		if input == "$ ls" {
			//list current dir
			for {
				scanner.Scan()
				nextInput := scanner.Text()
				fmt.Println(nextInput)

				if nextInput[0] == '$' {
					// No longer listing files/dirs
					break
				}
				splitInput := strings.Split(nextInput, " ")
				if splitInput[0] == "dir" {
					newNode := Node{
						text:     splitInput[1],
						size:     0,
						children: []*Node{},
						parent: curNode,
					}
					curNode.children = append(curNode.children, &newNode)

				} else {
					val, _ := strconv.Atoi(splitInput[0])
					newNode := Node{
						text:     splitInput[1],
						size:     val,
						children: []*Node{},
						parent: curNode,
					}
					curNode.children = append(curNode.children, &newNode)
				}
			}

		} else if input[0:4] == "$ cd" {
			splitInput := strings.Split(input, " ")
			if splitInput[2] == ".." {
				fmt.Println("Going to parent")
				fmt.Println(curNode.text)
				// Go to parent
				curNode = curNode.parent
			} else {
				// Go to child
				for _, child := range curNode.children {
					if child.text == splitInput[2] {
						curNode = child
						break
					}
				}
			}
		}
	}
	fmt.Println(SIZE_TO_CHECK)
	// fmt.Println(checkSubstr(string(input[0]), input, 0, 4)+1) // Part 1
	// fmt.Println(checkSubstr(string(input[0]), input, 0, 14)+1) // Part 2
}
